viam-cartographer/src/carto_facade/carto_facade.cc

```diff
@@ -554,8 +562,38 @@ void CartoFacade::GetPosition(viam_carto_get_position_response *r) {
     r->component_reference = bstrcpy(config.component_reference);
 };
 
-void CartoFacade::GetPointCloudMap(
-    viam_carto_get_point_cloud_map_response *r){};
+void CartoFacade::GetPointCloudMap(viam_carto_get_point_cloud_map_response *r) {
+    std::string pointcloud_map;
+    // Write or grab the latest pointcloud map in form of a string
+    std::shared_lock optimization_lock{optimization_shared_mutex,
+                                       std::defer_lock};
+    if (action_mode != ActionMode::LOCALIZING && optimization_lock.try_lock()) {
+        // We are able to lock the optimization_shared_mutex, which means
+        // that the optimization is not ongoing and we can grab the newest
+        // map
+        GetLatestSampledPointCloudMapString(pointcloud_map);
+        std::lock_guard<std::mutex> lk(viam_response_mutex);
+        latest_pointcloud_map = pointcloud_map;
+    } else {
+        // Either we are in localization mode or we couldn't lock the mutex
+        // which means the optimization process locked it and we need to use
+        // the backed up latest map
+        if (action_mode == ActionMode::LOCALIZING) {
+            LOG(INFO) << "In localization mode, using cached pointcloud map";
+        } else {
+            LOG(INFO)
+                << "Optimization is occuring, using cached pointcloud map";
+        }
+        std::lock_guard<std::mutex> lk(viam_response_mutex);
+        pointcloud_map = latest_pointcloud_map;
+    }
```

Comment:
We should make sure any changes from the "web UI does not keep loading map when localizing only" project gets copied in to CartoFacade

@PStrutz , @jeremyrhyde is there any code that should be moved here?

Reply:
I'll leave it to @PStrutz & @jeremyrhyde to reply, but from a cursory look, it appears as if @PStrutz 's current PR only touches go code: https://github.com/viamrobotics/viam-cartographer/pull/180/files